File: frontend/apps/ios/Pantopus/Features/Homes/ClaimOwnership/ClaimOwnershipSteps.swift
//
//  ClaimOwnershipSteps.swift
//  Pantopus
//
//  Step descriptors + form state for the claim-ownership wizard.
//

import Foundation

/// Steps the claim-ownership wizard can be on. Order is meaningful —
/// the wizard advances `start → upload → success` and back-navigates
/// `upload → start`. Success has no back chevron and ends the flow.
public enum ClaimOwnershipStep: String, CaseIterable, Sendable {
    case start
    case upload
    case success

    /// 1-based step number, nil for terminal step.
    public var stepNumber: Int? {
        switch self {
        case .start: 1
        case .upload: 2
        case .success: nil
        }
    }
}

/// Identifier for one of the two upload tiles.
public enum ClaimEvidenceSlot: String, CaseIterable, Sendable {
    case identity
    case ownership

    /// Backend `evidence_type` enum value sent on upload —
    /// `uploadEvidenceSchema` (`backend/routes/homeOwnership.js:43`).
    public var backendType: String {
        switch self {
        case .identity: "idv"
        case .ownership: "deed"
        }
    }

    public var title: String {
        switch self {
        case .identity: "Government ID"
        case .ownership: "Proof of ownership"
        }
    }

    public var acceptHint: String {
        "JPG, PNG, or PDF up to 10 MB"
    }
}

/// One picked file held in the VM until submit time.
public struct ClaimPickedFile: Sendable, Equatable {
    public let filename: String
    public let mimeType: String
    public let data: Data

    public init(filename: String, mimeType: String, data: Data) {
        self.filename = filename
        self.mimeType = mimeType
        self.data = data
    }

    public var sizeBytes: Int { data.count }
}

/// Per-slot upload state surfaced to the UI.
public enum ClaimSlotUiState: Sendable, Equatable {
    case empty
    case picked(file: ClaimPickedFile)
    case uploading(file: ClaimPickedFile, fraction: Double)
    case uploaded(file: ClaimPickedFile, fileURL: String)
    case failed(file: ClaimPickedFile, message: String)

    public var hasFile: Bool {
        switch self {
        case .empty: false
        default: true
        }
    }

    public var pickedFile: ClaimPickedFile? {
        switch self {
        case let .picked(file), let .uploading(file, _),
             let .uploaded(file, _), let .failed(file, _):
            file
        case .empty:
            nil
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/claim_ownership/ClaimOwnershipSteps.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.homes.claim_ownership

/**
 * Steps the claim-ownership wizard can be on. Order is meaningful — the
 * wizard advances `Start → Upload → Success` and back-navigates
 * `Upload → Start`. Success has no back chevron and ends the flow.
 *
 * Chrome (top-bar readout + progress fraction) is computed in the VM's
 * `computeChrome` rather than via per-step metadata on this enum; the
 * wizard doesn't survive process-death so there's nothing to (de)serialise.
 */
enum class ClaimOwnershipStep { Start, Upload, Success }

/**
 * Identifier for one of the two upload tiles. The `backendType` matches
 * the `evidence_type` Joi enum at `backend/routes/homeOwnership.js:43`.
 */
enum class ClaimEvidenceSlot(
    val backendType: String,
    val title: String,
) {
    Identity("idv", "Government ID"),
    Ownership("deed", "Proof of ownership"),
    ;

    val acceptHint: String get() = "JPG, PNG, or PDF up to 10 MB"
}

/** Per-slot upload state surfaced to the UI. */
sealed interface ClaimSlotState {
    data object Empty : ClaimSlotState

    data class Picked(val file: ClaimPickedFile) : ClaimSlotState

    data class Uploading(val file: ClaimPickedFile, val fraction: Float) : ClaimSlotState

    data class Uploaded(val file: ClaimPickedFile, val fileUrl: String) : ClaimSlotState

    data class Failed(val file: ClaimPickedFile, val message: String) : ClaimSlotState

    val hasFile: Boolean
        get() = this !is Empty

    val pickedFile: ClaimPickedFile?
        get() =
            when (this) {
                is Picked -> file
                is Uploading -> file
                is Uploaded -> file
                is Failed -> file
                is Empty -> null
            }
}

/** A single picked file buffered in memory until upload. */
data class ClaimPickedFile(
    val filename: String,
    val mimeType: String,
    val bytes: ByteArray,
) {
    val sizeBytes: Long get() = bytes.size.toLong()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ClaimPickedFile) return false
        if (filename != other.filename) return false
        if (mimeType != other.mimeType) return false
        if (!bytes.contentEquals(other.bytes)) return false
        return true
    }

    override fun hashCode(): Int {
        var result = filename.hashCode()
        result = 31 * result + mimeType.hashCode()
        result = 31 * result + bytes.contentHashCode()
        return result
    }
}

/** Outbound navigation events the screen consumes. */
sealed interface ClaimOwnershipOutboundEvent {
    /** Pop the wizard with no further navigation. */
    data object Dismiss : ClaimOwnershipOutboundEvent

    /** Pop the wizard and route to the user's claims list. */
    data object OpenClaimsList : ClaimOwnershipOutboundEvent
}
